File: mobile/src/main/java/mycroft/ai/BeaconActivity.java
package mycroft.ai;

import android.Manifest;
import android.annotation.TargetApi;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.UiThread;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import org.altbeacon.beacon.BeaconManager;

/**
 * Created by sarahkraynick on 2017-08-16.
 */
public class BeaconActivity extends AppCompatActivity {
    protected static final String TAG = "BeaconActivity";
    private static final int PERMISSION_REQUEST_COARSE_LOCATION = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Log.d(TAG, "onCreate");
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_beacon);
        verifyBluetooth();
        //logToDisplay("Application just launched");

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            // Android M Permission check
            if (this.checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                final AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setTitle(R.string.location_request_title);
                builder.setMessage(R.string.location_request_message);
                builder.setPositiveButton(android.R.string.ok, null);
                builder.setOnDismissListener(new DialogInterface.OnDismissListener() {

                    @TargetApi(23)
                    @Override
                    public void onDismiss(DialogInterface dialog) {
                        requestPermissions(new String[]{Manifest.permission.ACCESS_COARSE_LOCATION},
                                PERMISSION_REQUEST_COARSE_LOCATION);
                    }
                });
                builder.show();
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(@NonNull int requestCode,
                                           @NonNull String permissions[], @NonNull int[] grantResults) {
        switch (requestCode) {
            case PERMISSION_REQUEST_COARSE_LOCATION: {
                if (grantResults.length != 0) {
                    if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                        Log.d(TAG, "fine location permission granted");
                    } else {
                        final AlertDialog.Builder builder = new AlertDialog.Builder(this);
                        builder.setTitle(R.string.location_result_title);
                        builder.setMessage(R.string.location_result_message);
                        builder.setPositiveButton(android.R.string.ok, null);
                        builder.setOnDismissListener(new DialogInterface.OnDismissListener() {

                            @Override
                            public void onDismiss(DialogInterface dialog) {
                            }
                        });
                        builder.show();
                    }
                    return;
                }
            }
        }
    }

    public void onRangingClicked(View view) {
        Intent myIntent = new Intent(this, RangingActivity.class);
        this.startActivity(myIntent);
    }

    @Override
    public void onResume() {
        super.onResume();
        ((MycroftApplication) this.getApplicationContext()).setMonitoringActivity(this);
    }

    @Override
    public void onPause() {
        super.onPause();
        ((MycroftApplication) this.getApplicationContext()).setMonitoringActivity(null);
    }

    @UiThread
    private void verifyBluetooth() {
        try {
            if (!BeaconManager.getInstanceForApplication(this).checkAvailability()) {
                final AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setTitle(R.string.verify_bluetooth_not_enabled_title);
                builder.setMessage(R.string.verify_bluetooth_not_enabled_message);
                builder.setPositiveButton(android.R.string.ok, null);
                builder.setOnDismissListener(new DialogInterface.OnDismissListener() {
                    @Override
                    public void onDismiss(DialogInterface dialog) {
                        finish();
                        System.exit(0);
                    }
                });
                builder.show();
            }
        }
        catch (RuntimeException e) {
            final AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle(R.string.verify_bluetooth_not_enabled_title);
            builder.setMessage(R.string.verify_bluetooth_not_enabled_message);
            builder.setPositiveButton(android.R.string.ok, null);
            builder.setOnDismissListener(new DialogInterface.OnDismissListener() {

                @Override
                public void onDismiss(DialogInterface dialog) {
                    finish();
                }

            });
            builder.show();
        }
    }

    public void logToDisplay(final String line) {
        runOnUiThread(new Runnable() {
            public void run() {
                EditText editText = (EditText)BeaconActivity.this
                        .findViewById(R.id.monitoringText);
                editText.append(line+"\n");
            }
        });
    }
}

File: mobile/src/main/java/mycroft/ai/BeaconActivity.kt
package mycroft.ai

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.support.annotation.UiThread
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.widget.EditText
import kotlinx.android.synthetic.main.activity_beacon.*
import mycroft.ai.Constants.MycroftMobileConstants.PERMISSION_REQUEST_COARSE_LOCATION
import org.altbeacon.beacon.BeaconManager

/**
 * Created by sarahkraynick on 2017-08-16.
 */
class BeaconActivity : AppCompatActivity() {
    private var logTag = "BeaconActivity"

    override fun onCreate(savedInstanceState: Bundle?) {
        Log.d(logTag, "onCreate")
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_beacon)
        verifyBluetooth()

        startRangingButton.setOnClickListener { onRangingClicked() }

        // Android M Permission check
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M &&
                this.checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            val builder = AlertDialog.Builder(this)
            builder.setTitle(R.string.location_request_title)
            builder.setMessage(R.string.location_request_message)
            builder.setPositiveButton(android.R.string.ok, null)
            builder.setOnDismissListener {
                requestPermissions(arrayOf(Manifest.permission.ACCESS_COARSE_LOCATION),
                        PERMISSION_REQUEST_COARSE_LOCATION)
            }
            builder.show()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>, grantResults: IntArray) {
        when (requestCode) {
            PERMISSION_REQUEST_COARSE_LOCATION -> {
                if (grantResults.isNotEmpty() &&
                        grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    Log.d(logTag, "fine location permission granted")
                } else {
                    val builder = AlertDialog.Builder(this)
                    builder.setTitle(R.string.location_result_title)
                    builder.setMessage(R.string.location_result_message)
                    builder.setPositiveButton(android.R.string.ok, null)
                    builder.setOnDismissListener { }
                    builder.show()
                }
            }
        }
    }

    public override fun onResume() {
        super.onResume()
        (this.applicationContext as MycroftApplication).setBeaconActivity(this)
    }

    public override fun onPause() {
        super.onPause()
        (this.applicationContext as MycroftApplication).unsetBeaconActivity()
    }

    private fun onRangingClicked() {
        val myIntent = Intent(this, RangingActivity::class.java)
        this.startActivity(myIntent)
    }

    @UiThread
    private fun verifyBluetooth() {
        try {
            if (!BeaconManager.getInstanceForApplication(this).checkAvailability()) {
                val builder = AlertDialog.Builder(this)
                builder.setTitle(R.string.verify_bluetooth_not_enabled_title)
                builder.setMessage(R.string.verify_bluetooth_not_enabled_message)
                builder.setPositiveButton(android.R.string.ok, null)
                builder.setOnDismissListener { finish() }
                builder.show()
            }
        } catch (e: RuntimeException) {
            val builder = AlertDialog.Builder(this)
            builder.setTitle(R.string.verify_bluetooth_not_enabled_title)
            builder.setMessage(R.string.verify_bluetooth_not_enabled_message)
            builder.setPositiveButton(android.R.string.ok, null)
            builder.setOnDismissListener { finish() }
            builder.show()
        }

    }

    fun logToDisplay(line: String) {
        runOnUiThread {
            val editText = this@BeaconActivity
                    .findViewById(R.id.monitoringText) as EditText
            editText.append(line + "\n")
        }
    }
}
